Add EuroCrops dataset

favyen2 opened the pull request:
This adds the EuroCrops dataset (https://www.eurocrops.tum.de/index.html).

There are currently a few issues that I wasn't sure how to deal with

- The integer label does not directly correspond to a property of the vector features, but instead some mapping is needed. Currently I replaced `VectorDataset.label_name` with `label_fn: Callable[[fiona.Feature], int]` but we could do it another way instead.
- The vector labels are provided by year but VectorDataset does not seem to support timestamps. Should we add a way to get timestamp in VectorDataset, or wrap `__getitem__`?
- There are more than 255 classes in EuroCrops so `rasterio.features.rasterize` produces uint16 output instead of uint8. pytorch doesn't support uint16 so VectorDataset needs updates for this to work.
- EuroCrops has a class hierarchy (Level 3, 4, 5, 6) where higher levels are more fine-grained (see https://www.eurocrops.tum.de/taxonomy.html). Currently `EuroCrops(...)` accepts an `hcat_level` argument specifying the degree of granularity desired. Another option would be to return multiple masks, one for each level. A complicating factor is that some labels are not annotated at the leaves of the hierarchy but instead at intermediate nodes, presumably when the fine-grained label is not known, so if user requests Level 6 they may get some labels as "common_soft_wheat" and some as "winter_common_soft_wheat" (where "winter_common_soft_wheat" is a child of "common_soft_wheat" in the hierarchy). How do we want to deal with this? These labels could be leveraged by adding up the probabilities of the descendant categories when computing cross entropy loss, but I'm not sure if this training mode is or should be supported.

Here is example output in Portugal paired with Sentinel-2 imagery via intersection dataset:

![im1](https://github.com/microsoft/torchgeo/assets/90410805/a62515c0-149a-4ccd-8589-f3e7d4615024)
![im2](https://github.com/microsoft/torchgeo/assets/90410805/b15101a2-00c6-48ce-8bee-3bcac1a5fa19)

favyen2 commented on torchgeo/datasets/geo.py:
```diff
@@ -706,7 +706,16 @@ def __getitem__(self, query: BoundingBox) -> dict[str, Any]:
             masks = np.zeros((round(height), round(width)), dtype=np.uint8)
 
-        sample = {"mask": torch.tensor(masks), "crs": self.crs, "bbox": query}
+        # Convert masks from numpy to torch.
```
I realized that this functionality is already present in https://github.com/microsoft/torchgeo/blob/1c20678a1001805412595be9759dae18a5d956df/torchgeo/datasets/geo.py#L547. Should I move the code to a separate function?

adamjstewart replied:
Yes, this makes sense as a separate function. It can go in `torchgeo/datasets/utils.py`.

favyen2 commented:
@microsoft-github-policy-service agree company="Allen Institute for AI"

adamjstewart commented:
> The vector labels are provided by year but VectorDataset does not seem to support timestamps. Should we add a way to get timestamp in VectorDataset, or wrap `__getitem__`?

Assuming the year is found in the filename, this will be implemented by #1814

adamjstewart commented on torchgeo/datasets/eurocrops.py:
```diff
@@ -0,0 +1,285 @@
+
+"""EuroCrops dataset."""
+
+import csv
```
We also have a pandas dependency if you find that easier to work with, but csv is also fine for simple things.

adamjstewart commented on torchgeo/datasets/eurocrops.py:
```diff
@@ -0,0 +1,285 @@
+    publicly available self-declared crop reporting datasets from countries of the
+    European Union" into a unified format. The dataset is released under CC BY 4.0 Deed.
+
+    The dataset consists of shapefiles containing a total of X polygons. Each polygon
```
I assume X is a placeholder you plan on filling in

favyen2 replied:
Ah planned to fill it in before committing, fixed now!

adamjstewart commented on torchgeo/datasets/eurocrops.py:
```diff
@@ -0,0 +1,285 @@
+    ]
+
+    # Color palette to choose from.
+    # There are hundreds of classes so we pick color via modulo of class index.
```
Does the dataset offer any recommended cmap? I know CDL does but don't know much about EuroCrops

favyen2 replied:
The closest I found is this but it just visualizes the classes on a color scale:
https://github.com/maja601/EuroCrops/blob/main/webinar_demo/visualization.ipynb

adamjstewart replied:
Hmm, I don't love that either, but it's at least slightly more "official". Let's go with something similar to what they do. I'm afraid of cycling through a list because it means multiple classes will be mapped to the same color. The benefit of a floating point color scale is that it works even when the hcat level changes.

favyen2 replied:
OK I've switched to using continuous color map.

adamjstewart commented on torchgeo/datasets/eurocrops.py:
```diff
@@ -0,0 +1,285 @@
+        self,
+        paths: Union[str, Iterable[str]] = "data",
+        crs: Optional[CRS] = None,
+        res: float = 0.00001,
```
This may need to be changed depending on the units of the CRS. A default with a similar resolution as Sentinel-2 would make sense to me.

favyen2 replied:
The files range in CRS from WGS-84 to EPSG:31287 so I have set a default CRS (WGS-84), with res equivalent to 10 m/pixel at equator.

adamjstewart commented on torchgeo/datasets/eurocrops.py:
```diff
@@ -0,0 +1,285 @@
+        checksum: bool = False,
+        hcat_level: int = 6,
+    ) -> None:
+        """Initialize a new Dataset instance.
```
```suggestion
        """Initialize a new EuroCrops instance.
```

adamjstewart commented on torchgeo/datasets/eurocrops.py:
```diff
@@ -0,0 +1,285 @@
+        Raises:
+            DatasetNotFoundError: If dataset is not found and *download* is False.
+
+        .. versionchanged:: 0.5
+           *root* was renamed to *paths*.
```
You can remove this. This dataset was not added until 0.6, so the `root` parameter was never renamed for this dataset.

adamjstewart commented on torchgeo/datasets/eurocrops.py:
```diff
@@ -0,0 +1,285 @@
+
+        The classes are loaded corresponding to the hcat_level specified by the user.
+        """
+        assert isinstance(self.paths, str)
```
Seems like `self.paths` can never be a list of paths, is that correct? Not sure if it's possible to relax this constraint or if we should just change the type of the initial parameter.

favyen2 replied:
Yes it is similar to `cdl.py` I think.

adamjstewart commented on torchgeo/datasets/eurocrops.py:
```diff
@@ -0,0 +1,285 @@
+        Returns:
+            a matplotlib Figure with the rendered sample
+
+        .. versionchanged:: 0.3
+            Method now takes a sample dict, not a Tensor. Additionally, it is possible
+            to show subplot titles and/or use a custom suptitle.
```
You can remove this, the dataset didn't exist in version 0.3

adamjstewart commented on torchgeo/datasets/geo.py:
```diff
@@ -605,8 +605,8 @@ def __init__(
                 entry and returns a transformed version
-            label_name: name of the dataset property that has the label to be
-                rasterized into the mask
+            label_fn: function to extract the label to be rasterized into the mask from
+                the dataset Feature.
```
Will think about this. I'm not opposed, but want to make sure lambdas aren't too complicated for remote sensing researchers with limited Python skills.

favyen2 replied:
Let me know if this should be changed. Perhaps alternative where we have function like `VectorDataset.get_class_index(feat: fiona.Feature)` where default implementation returns `feat["properties"][self.label_name]` is better, then it can be overloaded to derive the feature differently without needing to overload the entire `__getitem__`.

adamjstewart replied:
Yeah, let's go with that alternative. Then, most datasets won't require this, we only need to think about it for particularly complicated datasets like EuroCrops.

adamjstewart reviewed (commented):
This looks great! Thanks for getting this done so quickly.

A couple remaining things:

1. Needs to be added to the documentation (the last 2 bullet points of https://torchgeo.readthedocs.io/en/stable/user/contributing.html#datasets)
2. We might need some way to select a subset of classes so we can train a model that predicts one of (maize, rice, soybean, wheat, other). This could also be done in the Lightning data module (next step). @yichiac have we done this yet for any of our other datasets? This is similar to the `classes` parameter of the CDL and NLCD datasets.

P.S. Curious to know how you're finding TorchGeo so far. VectorDataset is one of our underdeveloped features due to lack of example datasets, so the caveats you ran into are somewhat expected.

favyen2 commented:
> Needs to be added to the documentation (the last 2 bullet points of https://torchgeo.readthedocs.io/en/stable/user/contributing.html#datasets)

Thanks this helps a lot. I updated the relevant docs files.

> We might need some way to select a subset of classes so we can train a model that predicts one of (maize, rice, soybean, wheat, other).

I added a classes argument now. If a feature's class code isn't in the specified classes list, but is a descendant of one of the specified classes (based on the EuroCrops label hierarchy), then it will match to that closest ancestor.

> P.S. Curious to know how you're finding TorchGeo so far.

I've only been working in a small portion of the codebase (e.g. haven't tried training any models yet), but it's quite nice how convenient it is to intersect different datasets, and also to use one or more GeoTIFFs / shapefiles as labels. And the code is very high quality.

On the other hand I'm used to datasets that define windows / grid cells and then label them (like the [microsoft/satellite-imagery-labeling-tool](https://github.com/microsoft/satellite-imagery-labeling-tool) workflow I think). This way the set of inputs/outputs that model can be trained with is well-defined. I'm not sure how it would handle e.g. shapefiles that label all the points of some category in a city, but the bbox of the shapefile may include regions outside the city that are not labeled; you wouldn't want the detector to be trained on those regions as negative examples.

adamjstewart commented:
> I'm not sure how it would handle e.g. shapefiles that label all the points of some category in a city, but the bbox of the shapefile may include regions outside the city that are not labeled; you wouldn't want the detector to be trained on those regions as negative examples.

Each sampler accepts an `roi` as input that you could use to narrow down your region of interest to only the city.

adamjstewart commented:
With respect to the "year" info, how is that exposed in the dataset? Is it only found in the download URL or also in the filename? If the former, I have ideas. If the latter, this is now supported.

favyen2 commented:
> With respect to the "year" info, how is that exposed in the dataset? Is it only found in the download URL or also in the filename? If the former, I have ideas. If the latter, this is now supported.

Mm I saw the PR with VectorDataset date extraction you mentioned was merged, I added filename_regex/date_format so it can extract the year from the filename. I tried it out with some Sentinel-2 images and it created index and found the appropriate image correctly.

I need to update test to cover `_get_class_index`.

adamjstewart commented on torchgeo/datasets/geo.py:
```diff
@@ -725,13 +719,29 @@ def __getitem__(self, query: BoundingBox) -> dict[str, Any]:
 
         return sample
 
+    def _get_label(self, feature: Any) -> int:
```
Let's make this function public so it shows up in the docs. Also, shouldn't this be:
```suggestion
    def _get_label(self, feature: Feature) -> int:
```

favyen2 replied:
fiona.Feature doesn't exist in the min req version. I tried Union[Feature, dict[str, Any]] but it fails because the fiona.Feature doesn't exist. I tried to define Feature in a variable way (import fiona.Feature, but catch ImportError and define a placeholder) but it also doesn't work because the different definition of Feature would be caught as a re-definition.

favyen2 replied:
In the old fiona version, features are represented as dicts.

adamjstewart replied:
It looks like `fiona.Feature` was introduced in fiona 1.9, which is about a year old now. I'm okay with bumping the minimum supported version of fiona to 1.9 if that makes this easier to annotate. You would modify `pyproject.toml` and `requirements/min-reqs.old`.

Alternatively, you can do `"fiona.Feature"` to prevent evaluation during run-time.

adamjstewart replied:
To fix the RtD tests, you can add `fiona.Feature` to https://github.com/microsoft/torchgeo/blob/v0.5.1/docs/conf.py#L59 since it isn't actually documented.

favyen2 replied:
Ah ok perfect, I used `"fiona.Feature"`.

adamjstewart commented on torchgeo/datasets/eurocrops.py:
```diff
@@ -0,0 +1,266 @@
+
+    # Override variables to automatically extract timestamp.
+    filename_regex = r"""
+        ^(?P<prefix>[a-zA-Z]+(?:_[a-zA-Z]+)?)
```
Do the prefix and suffix have any meaning? Country code or something? Is there a webpage or file that documents this syntax?

adamjstewart replied:
I would personally be fine with:
```suggestion
        ^(?P<prefix>[A-Z_]+)
```
if that covers everything. I don't expect to find 4-digit sequences in the filename other than the year.

favyen2 replied:
> Do the prefix and suffix have any meaning? Country code or something? Is there a webpage or file that documents this syntax?

The EC21 stands for EuroCrops 2021 but some files just have EC instead.

It looks like the first part is either country or country + region. I updated the regex accordingly.

I didn't find any documentation about the filename format. But it should be "country{_region,}_YYYY_EC{21,}.shp".

adamjstewart commented on torchgeo/datasets/utils.py:
```diff
@@ -819,3 +819,29 @@ def path_is_vsi(path: str) -> bool:
+
+    Returns:
+        A :class:`torch.Tensor` with the same dtype as array unless array is uint16 or
+        uint32, in which case an int32 or int64 Tensor is returned, respectively.
```
Can you add:
```rst
.. versionadded:: 0.6
```
to the end? That will document which versions of TorchGeo support this.

adamjstewart commented on torchgeo/datasets/geo.py:
```diff
@@ -725,13 +719,29 @@ def __getitem__(self, query: BoundingBox) -> dict[str, Any]:
+            feature: the :class:`fiona.Feature` from which to extract the label.
+
+        Returns:
+            the integer label, or 0 if the feature should not be rendered.
```
Can you add:
```rst
.. versionadded:: 0.6
```

adamjstewart commented on torchgeo/datasets/eurocrops.py:
```diff
@@ -0,0 +1,267 @@
+    within the polygon in the year associated with the shapefile.
+
+    If you use this dataset in your research, please follow the citation guidelines at
+    https://github.com/maja601/EuroCrops#reference.
```
Add:
```rst
.. versionadded:: 0.6
```

favyen2 commented:
Thanks, made these updates!

I didn't understand why the docs build is failing, I just saw one warning:
> cannot import name 'LayerType' from 'timm.layers' (/home/docs/checkouts/readthedocs.org/user_builds/torchgeo/envs/1813/lib/python3.11/site-packages/timm/layers/__init__.py)

adamjstewart commented on torchgeo/datasets/eurocrops.py:
```diff
@@ -0,0 +1,270 @@
+        for fname, md5 in self.zenodo_files:
+            filepath = os.path.join(self.paths, fname)
+            if not check_integrity(filepath, md5 if self.checksum else None):
+                return False
```
This is the only remaining line not covered by unit tests

favyen2 replied:
Fixed!

adamjstewart commented:
Docs were failing due to #1824, I fixed this in #1827, updating your branch should solve the issue.

adamjstewart commented on docs/api/geo_datasets.csv:
```diff
@@ -9,6 +9,7 @@ Dataset,Type,Source,License,Size (px),Resolution (m)
 `EnviroAtlas`_,"Imagery, Masks","NAIP, NLCD, OpenStreetMap","CC-BY-4.0",-,1
 `Esri2020`_,Masks,Sentinel-2,"CC-BY-4.0",-,10
 `EU-DEM`_,Masks,"Aster, SRTM, Russian Topomaps","CSCDA-ESA",-,25
+`EuroCrops`_,Geometries,EU Countries,"CC-BY-SA-4.0",-,-
```
According to https://www.eurocrops.tum.de/index.html it should be:
```suggestion
`EuroCrops`_,Geometries,EU Countries,"CC-BY-4.0",-,-
```
right?

adamjstewart replied:
Ah, but the GitHub and Zenodo say CC-BY-SA-4.0. That's confusing. Let's go with what you have.

adamjstewart commented on torchgeo/datasets/eurocrops.py:
```diff
@@ -0,0 +1,270 @@
+
+    The `EuroCrops <https://www.eurocrops.tum.de/index.html>`__ dataset combines "all
+    publicly available self-declared crop reporting datasets from countries of the
+    European Union" into a unified format. The dataset is released under CC BY 4.0 Deed.
```
Here we say CC-BY-4.0 (because that's what the website says), but our docs say CC-BY-SA-4.0 (because that's what GitHub/Zenodo use). Not sure which to trust, but maybe we should be consistent.

adamjstewart reviewed (dismissed):
This looks awesome, great work! I'm running out of things to nitpick, so I think we can merge this soon. The license thing confuses me a bit. Let me reach out to the authors to clarify.

favyen2 commented on docs/conf.py:
```diff
@@ -67,6 +67,8 @@
     ("py:class", "torchvision.models._api.WeightsEnum"),
     ("py:class", "torchvision.models.resnet.ResNet"),
     ("py:class", "torchvision.models.swin_transformer.SwinTransformer"),
+    ("py:class", "fiona.Feature"),
+    ("py:class", "fiona.model.Feature"),
```
Since we don't refer to Feature in the codebase yet I changed our type annotations to use `fiona.model.Feature` which obviates the need for including `fiona.Feature` here. And it is alphabetical now.
